lib/Conversion/PolyToStandard/PolyToStandard.cpp

```diff
@@ -0,0 +1,156 @@
+  assert(inputs.size() == 1);
+  auto inputType = inputs[0].getType();
+  if (isa<PolynomialType>(inputType))
+    llvm_unreachable(
+        "A polynomial should never be the input to a materializeSource.");
+
+  if (!isa<TensorType>(inputType)) {
+    llvm::errs() << "materializeSource should only try to materialize tensor "
+                    "types back to "
+                    "polynomials, but found type: "
+                 << inputType;
+    llvm_unreachable("");
+  }
+
+  PolynomialType targetType = dyn_cast<PolynomialType>(inputType);
+  if (!targetType) llvm_unreachable("Target type must be a poly.poly");
+
+  return builder.create<PolyFromCoeffsOp>(loc, inputs[0], targetType.getRing());
+}
+
+static Value materializeTarget(OpBuilder &builder, Type type, ValueRange inputs,
+                               Location loc) {
+  assert(inputs.size() == 1);
+  auto inputType = inputs[0].getType();
+  if (!isa<PolynomialType>(inputType)) {
+    llvm::errs() << "materializeTarget should only try to materialize "
+                    "polynomial types to "
+                    "legal types, but found input: "
+                 << inputType;
+    llvm_unreachable("");
+  }
+
+  auto polyType = dyn_cast<PolynomialType>(inputType);
+  uint32_t idealDegree = polyType.getRing().ideal().getDegree();
+  auto startOp =
+      builder.create<arith::ConstantIntOp>(loc, 0, builder.getIndexType());
+  auto endOp = builder.create<arith::ConstantIntOp>(loc, idealDegree,
+                                                    builder.getIndexType());
+  return builder.create<ExtractSliceOp>(loc, inputs[0], startOp.getResult(),
+                                        endOp.getResult());
+}
+
+class PolyToStandardTypeConverter : public TypeConverter {
+ public:
+  PolyToStandardTypeConverter(MLIRContext *ctx) {
+    addConversion([](Type type) { return type; });
+    addConversion([ctx](PolynomialType type) -> Type {
+      RingAttr attr = type.getRing();
+      uint32_t idealDegree = attr.ideal().getDegree();
+      IntegerType elementTy =
+          IntegerType::get(ctx, attr.coefficientModulus().getBitWidth(),
+                           IntegerType::SignednessSemantics::Unsigned);
+      return RankedTensorType::get({idealDegree}, elementTy, attr);
+    });
+
+    addSourceMaterialization(materializeSource);
+    addArgumentMaterialization(materializeSource);
+    addTargetMaterialization(materializeTarget);
+  }
+};
+
+struct ConvertPolyFromCoeffs : public OpConversionPattern<PolyFromCoeffsOp> {
+  ConvertPolyFromCoeffs(mlir::MLIRContext *context)
+      : OpConversionPattern<PolyFromCoeffsOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      PolyFromCoeffsOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    // the input tensor replaces the op
+    rewriter.replaceAllUsesWith(op.getResult(), adaptor.getOperands()[0]);
+    rewriter.eraseOp(op);
```

Comment:
Your intuition is right, and I've been wrestling with this for a few days now. See https://reviews.llvm.org/D158107

The general advice in the docs is: any mutations to the IR just need to go through the ConversionPatternRerwiter, but not all the rewriter methods are handled properly (hence the bug with the diff above). It should just be rewriter.replaceOp().

The need for these conversions is I suppose a matter of minute details. If you do a full conversion, then the pass will fail on the first op it can't legalize (and if you call the entire dialect illegal then it would fail on these ops). If you do a partial conversion you can get by it, but then it will insert `unrealized_conversion_cast` ops which, at the end of the pass, are replaced with custom materializations if they can't be folded.

I admit I'm still learning this whole framework, but I think the most straightforward way to do it is to have a converter for every op and not include any custom materializations and do a full conversion. You would only worry about custom materializations if the pass gets very complicated (like bufferization) and you can then split it into multiple passes, each a partial conversion, with custom materializations to keep the IR consistent across the passes, which is the advice recommended in [this talk](https://drive.google.com/file/d/1VfVajitgf8ZPnd-HRkJvaJiFLhBsluXN/view)

I will go ahead and update the pattern above to use the correct rewriter method (I only learned about this discrepancy yesterday)